From 884e55d1abec1ae21f67d7684197e21a6b146c4a Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Tue, 25 Nov 2014 16:59:22 -0500
Subject: [net] ipv6: Do not treat a GSO_TCPV4 request from UDP tunnel over IPv6 as invalid

Message-id: <20141125165922.2907.38578.stgit@ahduyck-vm-rhel7>
Patchwork-id: 100949
O-Subject: [RHEL7.1 net PATCH 1/3] ipv6: Do not treat a GSO_TCPV4 request from UDP tunnel over IPv6 as invalid
Bugzilla: 1159577
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Bugzilla: 1159577
Brew Build: 8292276

Backport of upstream commit: b6fef4c6b8c1994ffe050dcdb9391427bf0d9882

 commit b6fef4c6b8c1994ffe050dcdb9391427bf0d9882
 Author: Alexander Duyck <alexander.h.duyck@redhat.com>
 Date:   Fri Nov 21 19:37:09 2014 -0800

    ipv6: Do not treat a GSO_TCPV4 request from UDP tunnel over IPv6 as invalid

    This patch adds SKB_GSO_TCPV4 to the list of supported GSO types handled by
    the IPv6 GSO offloads.  Without this change VXLAN tunnels running over IPv6
    do not currently handle IPv4 TCP TSO requests correctly and end up handing
    the non-segmented frame off to the device.

    Below is the before and after for a simple netperf TCP_STREAM test between
    two endpoints tunneling IPv4 over a VXLAN tunnel running on IPv6 on top of
    a 1Gb/s network adapter.

    Recv   Send    Send
    Socket Socket  Message  Elapsed
    Size   Size    Size     Time     Throughput
    bytes  bytes   bytes    secs.    10^6bits/sec

     87380  16384  16384    10.29       0.88      Before
     87380  16384  16384    10.03     895.69      After

    Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/ip6_offload.c b/net/ipv6/ip6_offload.c
index 1e45e56..0ff9301 100644
--- a/net/ipv6/ip6_offload.c
+++ b/net/ipv6/ip6_offload.c
@@ -95,7 +95,8 @@ static struct sk_buff *ipv6_gso_segment(struct sk_buff *skb,
  int nhoff;
 
  if (unlikely(skb_shinfo(skb)->gso_type &
-       ~(SKB_GSO_UDP |
+       ~(SKB_GSO_TCPV4 |
+         SKB_GSO_UDP |
          SKB_GSO_DODGY |
          SKB_GSO_TCP_ECN |
          SKB_GSO_GRE |
-- 
1.7.1